cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/CruiseControlReconciler.java

```diff
@@ -244,22 +252,33 @@ protected Future<Void> certificatesSecret(Clock clock) {
      */
     protected Future<Void> apiSecret() {
         if (cruiseControl != null) {
-            return secretOperator.getAsync(reconciliation.namespace(), CruiseControlResources.apiSecretName(reconciliation.name()))
+            if (topicOperatorEnabled) {
+                return secretOperator.getAsync(reconciliation.namespace(), CruiseControlResources.apiSecretName(reconciliation.name()))
                     .compose(oldSecret -> {
-                        Secret newSecret = cruiseControl.generateApiSecret(passwordGenerator);
-
-                        if (oldSecret != null)  {
-                            // The credentials should not change with every release
-                            // So if the secret with credentials already exists, we re-use the values
-                            // But we use the new secret to update labels etc. if needed
-                            newSecret.setData(oldSecret.getData());
-                        }
-                        
+                        return secretOperator.getAsync(reconciliation.namespace(), KafkaResources.entityTopicOperatorCcApiSecretName(reconciliation.name()))
+                            .compose(topicOperatorApiSecret -> {
+                                String cruiseControlAuthFile = oldSecret != null ? Util.decodeFromBase64(oldSecret.getData().get(API_AUTH_FILE_KEY)) : null;
+                                CruiseControl.AdminUser adminUser = topicOperatorApiSecret != null ? new CruiseControl.AdminUser(
+                                    Util.decodeFromBase64(topicOperatorApiSecret.getData().get(API_TO_ADMIN_NAME_KEY)),
+                                    Util.decodeFromBase64(topicOperatorApiSecret.getData().get(API_TO_ADMIN_PASSWORD_KEY))) : null;
+                                // generate a new CC API secret if there is no CC auth file, or there is no TO API secret, or TO admin password changed
+                                Secret newSecret = cruiseControlAuthFile == null || adminUser == null || !cruiseControlAuthFile.contains(adminUser.password())
+                                    ? cruiseControl.generateApiSecret(passwordGenerator, adminUser)
+                                    : cruiseControl.generateApiSecret(passwordGenerator, oldSecret);
```

Comment:
Is this really correct? Shouldn't you in one case consider both the TO Admin user and the existing CC users?

Reply:
I think it is fine. If there is existing CC users they also include the TO admin user, otherwise we generate a new CC API secret that includes the TO admin user.